dot/state/roottottrie.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/trie"
+)
+
+type rootToTrieMap struct {
+	rootToTrie map[common.Hash]*trie.Trie
+	sync.RWMutex
+}
+
+func newRootToTrieMap() *rootToTrieMap {
+	return &rootToTrieMap{
+		rootToTrie: make(map[common.Hash]*trie.Trie),
+	}
+}
+
+func (r *rootToTrieMap) get(root common.Hash) (t *trie.Trie, has bool) {
+	r.RLock()
+	defer r.RUnlock()
+	t, has = r.rootToTrie[root]
+	return t, has
+}
+
+func (r *rootToTrieMap) set(root common.Hash, trie *trie.Trie) {
+	r.Lock()
+	defer r.Unlock()
+	r.rootToTrie[root] = trie
+}
+
+func (r *rootToTrieMap) setIfUnset(root common.Hash, trie *trie.Trie) {
+	r.Lock()
+	defer r.Unlock()
+
+	_, exists := r.rootToTrie[root]
+	if exists {
+		return
+	}
+
+	r.rootToTrie[root] = trie
+}
```

Comment:
I would suggest to use the already written function because they use diferent locks for reading and writing
```suggestion
	_, exists := r.get(root)
	if exists { return }
	r.set(root, trie)
}
```

Reply:
It has to be atomic, this won't work (get is atomic, set is atomic but not both)